pyro/infer/mcmc/util.py

```diff
@@ -208,3 +212,131 @@ def log_prob(self, model_trace):
         log_probs = self._get_log_factors(model_trace)
         with shared_intermediates() as cache:
             return contract_to_tensor(log_probs, self._enum_dims, cache=cache)
+
+
+def _guess_max_plate_nesting(model, args, kwargs):
+    """
+    Guesses max_plate_nesting by running the model once
+    without enumeration. This optimistically assumes static model
+    structure.
+    """
+    with poutine.block():
+        model_trace = poutine.trace(model).get_trace(*args, **kwargs)
+    sites = [site for site in model_trace.nodes.values()
+             if site["type"] == "sample"]
+
+    dims = [frame.dim
+            for site in sites
+            for frame in site["cond_indep_stack"]
+            if frame.vectorized]
+    max_plate_nesting = -min(dims) if dims else 0
+    return max_plate_nesting
+
+
+def _pe_maker(model, model_args, model_kwargs, trace_prob_evaluator, transforms):
+    def potential_energy(params):
+        params_constrained = {k: transforms[k].inv(v) for k, v in params.items()}
+        cond_model = poutine.condition(model, params_constrained)
+        model_trace = poutine.trace(cond_model).get_trace(*model_args, **model_kwargs)
+        log_joint = -trace_prob_evaluator.log_prob(model_trace)
+        for name, t in transforms.items():
+            log_joint = log_joint + torch.sum(
+                t.log_abs_det_jacobian(params_constrained[name], params[name]))
+        return log_joint
+
+    return potential_energy
+
+
+def _get_init_params(model, model_args, model_kwargs, transforms, potential_fn, prototype_params,
+                     max_tries_initial_params=100):
+    params = prototype_params
+    for i in range(max_tries_initial_params):
+        potential_energy = potential_fn(params)
+        if not torch_isnan(potential_energy) and not torch_isinf(potential_energy):
+            return params
+        trace = poutine.trace(model).get_trace(*model_args, **model_kwargs)
+        samples = {name: trace.nodes[name]["value"].detach() for name in params}
```

Comment:
Do we need to `.detach()` here since the only place where we set/unset `requires_grad` is the verlet integrator?

Reply:
It seems that without it, there is an error happen. I'll check it against.

Reply:
Yes, it is required when our model has `pyro.param(...)` statements (so sample sites will be leaf variables, which make us unable to change `requires_grad` flag in velocity_verlet)

Reply:
I see. That because `pyro.param` also sets the `requires_grad` flag to True, I guess.

Reply:
I think so too.